### ISSUE TITLE
cannot find -lcuda

### ISSUE THREAD

**Issue by mengdeerer** [agentica-project/deepscaler#40]
During the startup of Deepscaler 1.5B model training, the linking process fails with an error indicating that the linker /usr/bin/ld cannot find the CUDA runtime library -lcuda. This issue prevents the training process from proceeding and strongly suggests a critical problem with the CUDA environment configuration.

> **Detailed Logs:**

(WorkerDict pid=415316) INFO 02-22 13:06:42 model_runner_base.py:120] Writing input of failed execution to /tmp/err_execute_model_input_20250222-130642.pkl...
(WorkerDict pid=415316) /usr/bin/ld: skipping incompatible /lib/i386-linux-gnu/libcuda.so when searching for -lcuda
(WorkerDict pid=415316) /usr/bin/ld: cannot find -lcuda
(WorkerDict pid=415316) collect2: error: ld returned 1 exit status

> **steps to reproduce**

Run the training script scripts/ablations/run_deepscaler_1.5b_2k.sh (or your specific training script).
Observe the log output. The error /usr/bin/ld: cannot find -lcuda appears during the model loading or initialization phase.

> **python packages**

accelerate                        1.4.0
aiohappyeyeballs                  2.4.6
aiohttp                           3.11.12
aiosignal                         1.3.2
annotated-types                   0.7.0
antlr4-python3-runtime            4.7.2
anyio                             4.8.0
asttokens                         3.0.0
async-timeout                     5.0.1
attrs                             25.1.0
blobfile                          3.0.0
cachetools                        5.5.2
certifi                           2025.1.31
charset-normalizer                3.4.1
click                             8.1.8
cloudpickle                       3.1.1
codetiming                        1.4.0
comm                              0.2.2
contourpy                         1.3.1
cycler                            0.12.1
datasets                          3.3.2
debugpy                           1.8.12
decorator                         5.1.1
deepscaler                        0.0.0         deepscaler
dill                              0.3.8
diskcache                         5.6.3
distro                            1.9.0
docker-pycreds                    0.4.0
docstring_parser                  0.16
einops                            0.8.1
exceptiongroup                    1.2.2
executing                         2.1.0
fastapi                           0.115.8
filelock                          3.17.0
flash_attn                        2.7.3
fonttools                         4.56.0
frozenlist                        1.5.0
fsspec                            2024.12.0
gguf                              0.10.0
gitdb                             4.0.12
GitPython                         3.1.44
google-api-core                   2.24.1
google-auth                       2.38.0
google-cloud-aiplatform           1.81.0
google-cloud-bigquery             3.29.0
google-cloud-core                 2.4.2
google-cloud-resource-manager     1.14.1
google-cloud-storage              2.19.0
google-crc32c                     1.6.0
google-resumable-media            2.7.2
googleapis-common-protos          1.68.0
grpc-google-iam-v1                0.14.0
grpcio                            1.70.0
grpcio-status                     1.70.0
h11                               0.14.0
httpcore                          1.0.7
httptools                         0.6.4
httpx                             0.28.1
huggingface-hub                   0.29.1
hydra-core                        1.3.2
idna                              3.10
importlib_metadata                8.6.1
interegular                       0.3.3
ipykernel                         6.29.5
ipython                           8.32.0
jedi                              0.19.2
Jinja2                            3.1.5
jiter                             0.8.2
joblib                            1.4.2
jsonschema                        4.23.0
jsonschema-specifications         2024.10.1
jupyter_client                    8.6.3
jupyter_core                      5.7.2
kiwisolver                        1.4.8
lark                              1.2.2
latex2sympy2                      1.9.1
llvmlite                          0.44.0
lm-format-enforcer                0.10.6
lxml                              5.3.0
MarkupSafe                        3.0.2
matplotlib                        3.10.0
matplotlib-inline                 0.1.7
mistral_common                    1.5.3
mpmath                            1.3.0
msgpack                           1.1.0
msgspec                           0.19.0
multidict                         6.1.0
multiprocess                      0.70.16
nest_asyncio                      1.6.0
networkx                          3.4.2
ninja                             1.11.1.3
numba                             0.61.0
numpy                             1.26.4
nvidia-cublas-cu12                12.1.3.1
nvidia-cuda-cupti-cu12            12.1.105
nvidia-cuda-nvrtc-cu12            12.1.105
nvidia-cuda-runtime-cu12          12.1.105
nvidia-cudnn-cu12                 9.1.0.70
nvidia-cufft-cu12                 11.0.2.54
nvidia-curand-cu12                10.3.2.106
nvidia-cusolver-cu12              11.4.5.107
nvidia-cusparse-cu12              12.1.0.106
nvidia-ml-py                      12.570.86
nvidia-nccl-cu12                  2.20.5
nvidia-nvjitlink-cu12             12.8.61
nvidia-nvtx-cu12                  12.1.105
omegaconf                         2.3.0
openai                            1.63.2
opencv-python-headless            4.11.0.86
orjson                            3.10.15
outlines                          0.0.46
packaging                         24.2
pandas                            2.2.3
parso                             0.8.4
partial-json-parser               0.2.1.1.post5
pexpect                           4.9.0
pickleshare                       0.7.5
pillow                            11.1.0
pip                               25.0
platformdirs                      4.3.6
prometheus_client                 0.21.1
prometheus-fastapi-instrumentator 7.0.2
prompt_toolkit                    3.0.50
propcache                         0.3.0
proto-plus                        1.26.0
protobuf                          5.29.3
psutil                            6.1.1
ptyprocess                        0.7.0
pure_eval                         0.2.3
py-cpuinfo                        9.0.0
pyairports                        2.1.1
pyarrow                           19.0.1
pyasn1                            0.6.1
pyasn1_modules                    0.4.1
pybind11                          2.13.6
pycountry                         24.6.1
pycryptodomex                     3.21.0
pydantic                          2.10.6
pydantic_core                     2.27.2
Pygments                          2.19.1
pylatexenc                        2.10
pyparsing                         3.2.1
python-dateutil                   2.9.0.post0
python-dotenv                     1.0.1
pytz                              2025.1
PyYAML                            6.0.2
pyzmq                             26.2.1
ray                               2.42.1
referencing                       0.36.2
regex                             2024.11.6
requests                          2.32.3
rpds-py                           0.23.1
rsa                               4.9
safetensors                       0.5.2
scikit-learn                      1.6.1
scipy                             1.15.2
sentence-transformers             3.4.1
sentencepiece                     0.2.0
sentry-sdk                        2.22.0
setproctitle                      1.3.4
setuptools                        75.8.0
shapely                           2.0.7
six                               1.17.0
smmap                             5.0.2
sniffio                           1.3.1
stack_data                        0.6.3
starlette                         0.45.3
sympy                             1.13.3
tabulate                          0.9.0
tensordict                        0.5.0
threadpoolctl                     3.5.0
tiktoken                          0.8.0
timm                              1.0.14
tokenizers                        0.21.0
torch                             2.4.0
torchvision                       0.19.0
tornado                           6.4.2
tqdm                              4.67.1
traitlets                         5.14.3
transformers                      4.47.1
triton                            3.0.0
typing_extensions                 4.12.2
tzdata                            2025.1
urllib3                           2.3.0
uvicorn                           0.34.0
uvloop                            0.21.0
verl                              0.1           deepscaler/verl
vllm                              0.6.3
wandb                             0.19.7
watchfiles                        1.0.4
wcwidth                           0.2.13
websockets                        15.0
wheel                             0.45.1
xformers                          0.0.27.post2
xxhash                            3.5.0
yarl                              1.18.3
zipp                              3.21.0

> **cuda environments**

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Jan__6_16:45:21_PST_2023
Cuda compilation tools, release 12.0, V12.0.140
Build cuda_12.0.r12.0/compiler.32267302_0

> **torch输出**

>>> import torch
>>> print(torch.version.cuda)
12.1

**Comment by lambda7xx** [comment]
you can try to set your cuda  path in ```~/.bashrc```. We require cuda-12

```
export PATH=/usr/local/cuda/bin:$PATH
export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH
```

**Comment by mengdeerer** [comment]
thank you for your suggestion, i have tried this command "export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH" ,and what in this directionary is a libcudda.so file which have a symbolic link to /usr/lib/x86_64-linux-gnu/libcuda.so.535.183.01, but there is nothing changed, **the error still exists**

**Comment by mengdeerer** [comment]
After updating CUDA version from 12.0 to 12.4 and reinstalling all packages, the problem was solved. Thanks  😄
